common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -0,0 +1,255 @@
+	) {
+		return
+	}
+
+	close(s.shutdownChan)
+
+	s.processor.Stop()
+
+	s.rescheduleTasks()
+
+	s.logger.Info("interleaved weighted round robin task scheduler stopped")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Submit(
+	task PriorityTask,
+
+) {
+	channel := s.getOrCreateTaskChannel(s.priorityToWeight[task.GetPriority()])
+	channel.Chan() <- task
+	s.notifyDispatcher()
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) eventLoop() {
+	for {
+		select {
+		case <-s.notifyChan:
+			for s.hasRemainingTasks() {
+				weightedChannels := s.channels()
+				s.dispatchTasks(weightedChannels)
+			}
+
+		case <-s.shutdownChan:
+			return
+		}
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) getOrCreateTaskChannel(
+	weight int,
+) *WeightedChannel {
+	s.RLock()
+	channel, ok := s.weightToTaskChannels[weight]
+	if ok {
+		s.RUnlock()
+		return channel
+	}
+	s.RUnlock()
+
+	s.Lock()
+	defer s.Unlock()
+
+	channel, ok = s.weightToTaskChannels[weight]
+	if ok {
+		return channel
+	}
+
+	channel = NewWeightedChannel(weight, WeightedChannelDefaultSize)
+	s.weightToTaskChannels[weight] = channel
+
+	weightedChannels := make(WeightedChannels, 0, len(s.weightToTaskChannels))
+	for _, weightedChan := range s.weightToTaskChannels {
+		weightedChannels = append(weightedChannels, weightedChan)
+	}
+	sort.Sort(weightedChannels)
+
+	iwrrChannels := make([]*WeightedChannel, 0, len(weightedChannels))
+	maxWeight := weightedChannels[len(weightedChannels)-1].Weight()
+	for round := maxWeight - 1; round > -1; round-- {
+		for index := len(weightedChannels) - 1; index > -1 && weightedChannels[index].Weight() > round; index-- {
+			iwrrChannels = append(iwrrChannels, weightedChannels[index])
+		}
+	}
+	s.iwrrChannels = iwrrChannels
+
+	return channel
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) channels() []*WeightedChannel {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.iwrrChannels
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) notifyDispatcher() {
+	if s.isStopped() {
+		s.rescheduleTasks()
+		return
+	}
+
+	select {
+	case s.notifyChan <- struct{}{}:
+	default:
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) dispatchTasks(
+	channels []*WeightedChannel,
+) {
+
+LoopDispatch:
+	for _, channel := range channels {
+		select {
+		case task := <-channel.Chan():
+			s.processor.Submit(task)
+
+		case <-s.shutdownChan:
+			return
+
+		default:
+			continue LoopDispatch
+		}
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) hasRemainingTasks() bool {
+	s.RLock()
+	defer s.RUnlock()
+
+	for _, weightedChan := range s.weightToTaskChannels {
+		if weightedChan.Len() > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) rescheduleTasks() {
+	s.RLock()
+	defer s.RUnlock()
+
+DrainLoop:
+	for _, channel := range s.weightToTaskChannels {
+		for {
+			select {
+			case task := <-channel.Chan():
+				task.Reschedule()
+
+			default:
+				continue DrainLoop
```

Comment:
doesn't this loop need an exit condition?  Am I missing something?  Or is this only usable in an entire-process-lifetime context?

Reply:
this function will only be invoked when scheduler is shutdown.